File: Stack/CPP/dynamicArray_Implementation.cpp
#include<iostream>
#include<vector>
using namespace std;

template<typename T>
class CustomStack{
	    vector<T> data;
        int index;
    public:

        CustomStack(){  
          this->index = -1;
        }

        CustomStack(const CustomStack& other) {
          data = other.data;  
          index = other.index;
        }

        bool isempty(){
          return index == -1;
        }

        T top(){
          try {
             if (isempty()) {
                  throw runtime_error("Error: Stack UnderFlow");
              }
              return data[index];
          }catch (const exception& e) {
              cout << e.what()<<" -- Default Value : ";;
              return T();
          }
        }

        void push(T val){
          index++;
          data.push_back(val);
        }

        void pop(){
        if(index == -1){
           cout<<"Stack UnderFlow"<<endl;
           return;
          }
        index--;
        }

        ~CustomStack() {
        // No need
      }

};


int main(){
    CustomStack<long long> stk;
    stk.pop(); 
    cout<<stk.isempty()<<endl;
    stk.push(7);
    cout<<stk.isempty()<<endl;
    cout<<stk.top()<<endl;
    stk.pop();
    cout<<stk.top()<<endl;
    cout<<stk.isempty()<<endl;
    stk.push(6);
    cout<<"---------------------"<<endl;
    CustomStack<long long > obj(stk);
    cout<<obj.top()<<endl;
}

File: Stack/PYTHON/dynamicArray_Implementation.py
class CustomStack:
    def __init__(self):
        self.data = []
        self.index = -1

    def isempty(self):
        return self.index == -1

    def top(self):
        if self.isempty():
            print("Warning: Stack is Empty")
            return None
        return self.data[self.index]

    def push(self, val):
        self.index += 1
        self.data.append(val)

    def pop(self):
        if self.index == -1:
            print("Stack Is Empty")
            return
        self.index -= 1



stk = CustomStack()
stk.pop()
print(stk.isempty())
stk.push(6)
print(stk.isempty())
print(stk.top())
stk.pop()
print(stk.top())
print(stk.isempty())


